### PR TITLE
fix(dist): correct local-rank/device detection under ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE

### PR DESCRIPTION
**Pull request by saforem2**
## Problem

Running `ezpz launch` with `ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE` on Sunspot/Aurora crashed:

```
RuntimeError: The device index is out of range. It must be in [0, 6), but got 11.
```

Repro command (24 ranks, 6 per node, COMPOSITE collapses 12 tiles → 6 composite GPUs):

```bash
ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE ezpz launch -n 24 -ppn 6 \
  python3 -m ezpz.examples.fsdp_tp --model=agpt-2b ... --tp=1 --dp-replicate=24 ...
```

## Root cause

Two independent defects combined:

1. **Unrecognized launcher env var.** `get_local_rank()` checked `LOCAL_RANK, PMI_LOCAL_RANK, OMPI_…, MPI_LOCALRANKID, MPICH_LOCALRANKID, SLURM_LOCAL_ID` — but the cray-pals `mpiexec` used on Aurora/Sunspot sets **`PALS_LOCAL_RANKID`** / **`PMIX_LOCAL_RANK`**, neither of which was in the list. Confirmed from the live system's `libpals.so.0.0.0` (`strings`) and ezpz's own `affinity.sh` (which reads `PMIX_LOCAL_RANK`). So it fell through to the modulo fallback.

2. **Device-count modulo with a stale GPU count.** The fallback was `get_rank() % get_gpus_per_node()`, and `get_gpus_per_node()` prefers the shell-exported `NGPU_PER_HOST=12` (a **FLAT-mode** tile constant hardcoded in `utils.sh`/`ezpz.sh`). Under COMPOSITE only 6 composite GPUs exist, so rank 23 computed `23 % 12 = 11` → `torch.xpu.set_device(11)` → out of range `[0, 6)`.

Confirmed by the workaround `NGPU_PER_HOST=6 ezpz launch ...` fixing it — which changed the *fallback* result (`23 % 6 = 5`), proving the fallback path was being taken.

## Fix

- **Recognize the real launcher vars.** Add `PALS_LOCAL_RANKID` and `PMIX_LOCAL_RANK` to `get_local_rank()`'s list, so the launcher's authoritative local rank is used in every FLAT/COMPOSITE/`-ppn` configuration (this alone fixes the crash; the fallback is no longer reached on these systems).
- **Fix the fallback math.** Use ranks-per-node (`world_size // num_nodes`) instead of the device count, so the result stays in `[0, ppn)` whenever ranks-per-node ≠ device-count — this also fixes plain `-ppn N` under-subscription on any node, independent of COMPOSITE. Degrades to the historical device-count behavior when nnodes can't be determined; guarded against `ZeroDivisionError`.
- **Reconcile `get_gpus_per_node()` with reality.** Clamp a stale FLAT env hint *down* to the visible XPU device count (via a new login-node-safe `_xpu_device_count_safe()`), so `get_node_index()` and the per-host summary line stop over-counting under COMPOSITE. Clamping is **one-directional** (never bumps a deliberate under-subscription up) and **XPU-only** (COMPOSITE has no CUDA analogue) — CUDA behavior is byte-identical.

## Tests

`tests/test_distributed.py` (+10 tests, 199 pass):
- `get_local_rank()` reads `PALS_LOCAL_RANKID` / `PMIX_LOCAL_RANK`; PALS var beats the modulo fallback under a stale `NGPU_PER_HOST=12`.
- Fallback uses ranks-per-node: rank 23 of `-n 24 -ppn 6` → local 5 (not 11); CPU single-node yields distinct ranks; indeterminate topology → 0 (no crash).
- `get_gpus_per_node()` clamps `12 → 6` under COMPOSITE; leaves FLAT (`12 == 12`), under-subscription (`4`), and CUDA hints untouched; survives a login-node XPU probe error.

Full suite: 1264 passed, 24 skipped (1 pre-existing unrelated failure on this box: `command_exists("python")` — no bare `python` on PATH, fails identically on `main`).

## Validation status

Root cause confirmed empirically at every layer (libpals strings on the live node, `affinity.sh`, the `NGPU_PER_HOST=6` workaround, and the crash's own `[0, 6)` bound proving `xpu.device_count()==6` under COMPOSITE). On-node re-run of the original command is a recommended follow-up but not required to prove correctness — the fix removes both the wrong-var and wrong-math paths deterministically.

`release:patch`

## Summary by Sourcery

Fix local rank and GPU-per-node detection to prevent out-of-range device indices under Intel XPU COMPOSITE topology and other under-subscribed configurations.

Bug Fixes:
- Recognize PALS_LOCAL_RANKID and PMIX_LOCAL_RANK environment variables so get_local_rank uses the launcher-provided local rank on Aurora/Sunspot instead of falling back to incorrect modulo logic.
- Change the local-rank fallback to compute rank modulo ranks-per-node (derived from world size and node count) and guard against indeterminate topologies to avoid ZeroDivisionError and out-of-range device indices.
- Clamp stale NGPU_PER_HOST values down to the visible Intel XPU device count under COMPOSITE while preserving intentional under-subscription and CUDA behavior.

Tests:
- Add comprehensive distributed tests covering new local-rank env vars, ranks-per-node fallback behavior across GPU and CPU-only scenarios, and XPU-specific clamping of NGPU_PER_HOST including error and CUDA edge cases.

### CODE REVIEW

File: tests/test_distributed.py [hunk 1]
```diff
@@ -411,6 +538,71 @@ def test_cpu_only(self, monkeypatch):
...
+            assert dist.get_gpus_per_node() == 6
+
+    def test_flat_no_clamp_when_env_matches_devices(self, monkeypatch):
+        """FLAT mode: ``NGPU_PER_HOST=12`` == device count ⇒ unchanged."""
+        monkeypatch.setenv("NGPU_PER_HOST", "12")
+        with (
+            patch.object(torch.xpu, "is_available", return_value=True),
+            patch.object(torch.xpu, "device_count", return_value=12),
+        ):
+            assert dist.get_gpus_per_node() == 12
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_distributed.py:565]: Good catch — added `test_no_xpu_attribute_keeps_env_hint` in `4ba94fb`. It `delattr`s `torch.xpu` and asserts `get_gpus_per_node()` returns the env hint verbatim, exercising the `hasattr(torch, "xpu")` guard in `_xpu_device_count_safe` (which the helper already uses, so no source change was needed). 200 pass in `test_distributed.py`.

File: src/ezpz/distributed.py [hunk 2]
```diff
@@ -296,18 +322,49 @@ def get_gpus_per_node() -> int:
     ):
         val = os.environ.get(var)
         if val is not None and val != "":
-            return int(val)
+            env_hint = int(val)
+            break
+
+    xpu_count = _xpu_device_count_safe()
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/distributed.py]: Confirmed and fixed in `c58c84f` — good catch, this was a real regression. I reproduced it against `origin/main` (returns `12`) vs. this branch (raised `ModuleNotFoundError`) with `NGPU_PER_HOST` set and torch hidden from the import machinery.

Root cause: the COMPOSITE clamp moved the `import torch` (via `_xpu_device_count_safe`) ahead of the env-hint return, so the torch-less launch/PBS topology path (`pbs._infer_topology` → `get_gpus_per_node`) crashed instead of returning the hint.

Fix: `_xpu_device_count_safe()` now probes torch only if it's **already imported** (`sys.modules.get("torch")`), never forcing an import. Torch-less callers get the env hint back verbatim (no import, no crash); worker processes — where `setup_torch` has already imported torch — still get the COMPOSITE clamp (`NGPU_PER_HOST=12` → `xpu.device_count()=6`). Regression test: `test_env_hint_honored_without_torch`.

File: src/ezpz/distributed.py [hunk 3]
```diff
@@ -296,18 +322,49 @@ def get_gpus_per_node() -> int:
...
+    import torch
+
+    if not hasattr(torch, "xpu"):
+        return None
+    try:
+        if torch.xpu.is_available():
+            return torch.xpu.device_count()
+    except Exception:
+        return None
+    return None
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/distributed.py]: Fixed in `c58c84f` — you're right on both counts (the torch-less break and the heavy import on the common path). `_xpu_device_count_safe()` no longer forces `import torch`; it reads `sys.modules.get("torch")` and returns `None` if torch isn't already loaded. So the env-hint fast path takes no import at all, and torch-less entrypoints (`ezpz launch` / `pbs._infer_topology`) return the env hint as before. The COMPOSITE clamp still fires in worker processes where torch is already imported. Verified with `test_env_hint_honored_without_torch` plus the existing clamp test.
